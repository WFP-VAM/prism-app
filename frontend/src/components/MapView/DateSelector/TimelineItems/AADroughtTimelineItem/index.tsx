import { createStyles, makeStyles } from '@material-ui/core';
import React, { memo, useMemo } from 'react';
import 'react-datepicker/dist/react-datepicker.css';
import { DateItem, DateRangeType } from 'config/types';
import { datesAreEqualWithoutTime } from 'utils/date-utils';

const AADroughtTimelineItem = memo(
  ({
    concatenatedLayers,
    currentDate,
    dateItemStyling,
    isDateAvailable,
  }: AADroughtTimelineItemProps) => {
    // Pre-compute the matching indices for all layers
    const classes = useStyles();

    const displayDateMatches = useMemo(
      () =>
        concatenatedLayers.map(layerDates =>
          layerDates.findIndex(i =>
            datesAreEqualWithoutTime(i.displayDate, currentDate.value),
          ),
        ),
      [concatenatedLayers, currentDate.value],
    );

    const layerMatches = useMemo(() => {
      const queryDateMatches = concatenatedLayers.map(layerDates =>
        layerDates.findIndex(i =>
          datesAreEqualWithoutTime(i.queryDate, currentDate.value),
        ),
      );

      return displayDateMatches.map((displayDateMatch, layerIndex) =>
        queryDateMatches[layerIndex] > -1 &&
        !datesAreEqualWithoutTime(
          concatenatedLayers[layerIndex][displayDateMatch].queryDate,
          currentDate.value,
        )
          ? queryDateMatches[layerIndex]
          : displayDateMatch,
      );
    }, [concatenatedLayers, currentDate.value, displayDateMatches]);

    const hasNextItemDirectionForward = (
      _matchingDate: DateItem,
      _layerDates: DateItem[],
    ): boolean => false;

    const hasNextItemDirectionBackward = (
      _matchingDate: DateItem,
      _layerDates: DateItem[],
    ): boolean => false;

    const isQueryDate = (date: DateItem): boolean =>
      datesAreEqualWithoutTime(date.queryDate, date.displayDate);

    return (
      <>
        {/* Add a small grey line to indicate where dates are overlapping */}
        {layerMatches.length >= 1 && isDateAvailable && (
          <div
            className={dateItemStyling[3].class}
            style={{
              height: 4,
              // TODO - handle more than 3 layers
              top: 10 * Math.min((layerMatches?.length || 0) + 1, 3),
            }}
            key={Math.random()}
            role="presentation"
          />
        )}
        {layerMatches.map((idx, layerIndex) => {
          const layerDates = concatenatedLayers[layerIndex];
          const matchingDateItemInLayer: DateItem | undefined =
            idx > -1 ? layerDates[idx] : undefined;

          if (!matchingDateItemInLayer) {
            return null;
          }

          return (
            <React.Fragment key={Math.random()}>
              {/* Add a directional arrow forward if previous item is a start date */}
              {hasNextItemDirectionForward(
                matchingDateItemInLayer,
                layerDates,
              ) && (
                <div
                  className={`${dateItemStyling[layerIndex].layerDirectionClass} ${classes.layerDirectionBase}`}
                />
              )}

              {/* Add a directional arrow backward if next item is an end date */}
              {hasNextItemDirectionBackward(
                matchingDateItemInLayer,
                layerDates,
              ) && (
                <div
                  className={`${dateItemStyling[layerIndex].layerDirectionClass} ${classes.layerDirectionBase} ${classes.layerDirectionBackwardBase}`}
                />
              )}

              {/* Add a bold square if queryDate (emphasis), normal otherwise */}
              <div
                className={`${
                  isQueryDate(matchingDateItemInLayer)
                    ? dateItemStyling[layerIndex].emphasis
                    : dateItemStyling[layerIndex].class
                }`}
                role="presentation"
              />
            </React.Fragment>
          );
        })}
      </>
    );
  },
);

const useStyles = makeStyles(() =>
  createStyles({
    layerDirectionBase: {
      display: 'block',
      position: 'absolute',
      borderTop: '5px solid transparent',
      borderBottom: '5px solid transparent',
      height: '0px',
      zIndex: 1,
      left: 0,
      pointerEvents: 'none',
    },

    layerDirectionBackwardBase: {
      right: 0,
      transform: 'rotate(180deg)',
    },
  }),
);

export interface AADroughtTimelineItemProps {
  concatenatedLayers: DateItem[][];
  currentDate: DateRangeType;
  dateItemStyling: {
    class: string;
    color: string;
    layerDirectionClass?: string;
    emphasis?: string;
  }[];
  isDateAvailable: boolean;
}

export default AADroughtTimelineItem;
