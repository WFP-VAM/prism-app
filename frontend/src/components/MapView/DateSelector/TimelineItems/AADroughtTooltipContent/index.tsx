import { DateRangeType } from 'config/types';
import { compact } from 'lodash';
import React, { FC } from 'react';
import { datesAreEqualWithoutTime, getFormattedDate } from 'utils/date-utils';
import { useSafeTranslation } from 'i18n';
import TooltipItem from '../TooltipItem';
import { DateCompatibleLayerWithDateItems } from '../../utils';
import { DateItemStyle } from '../types';

const AAdroughtTooltipContent: FC<AAdroughtTooltipContentProps> = ({
  date,
  orderedLayers,
  dateItemStyling,
}) => {
  const { t } = useSafeTranslation();

  const tooltipTitleArray: React.JSX.Element[] = compact(
    orderedLayers.map((selectedLayer, layerIndex) => {
      // find closest date element for layer
      const dateItem = selectedLayer.dateItems.find(item =>
        datesAreEqualWithoutTime(item.displayDate, date.value),
      );
      if (!dateItem) {
        return null;
      }

      // Display range dates when available
      const formattedDate =
        dateItem.startDate && dateItem.endDate
          ? `${getFormattedDate(
              dateItem.startDate,
              'monthDay',
            )} - ${getFormattedDate(dateItem.endDate, 'monthDay')}`
          : getFormattedDate(dateItem.queryDate, 'monthDay');
      return (
        <TooltipItem
          key={`Tootlip-${date.label}-${date.value}-${selectedLayer.title}`}
          layerTitle={`${t(selectedLayer.title)}:  ${formattedDate}`}
          color={dateItemStyling[layerIndex].color}
        />
      );
    }),
  );
  // eslint-disable-next-line fp/no-mutating-methods
  tooltipTitleArray.unshift(<div key={date.label}>{date.label}</div>);
  return tooltipTitleArray;
};

interface AAdroughtTooltipContentProps {
  date: DateRangeType;
  orderedLayers: DateCompatibleLayerWithDateItems[];
  dateItemStyling: DateItemStyle[];
}

export default AAdroughtTooltipContent;
