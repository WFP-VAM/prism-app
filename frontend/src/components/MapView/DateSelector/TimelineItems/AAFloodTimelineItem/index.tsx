import { useSelector } from 'react-redux';
import { useState } from 'react';
import { RootState } from 'context/store';
import { DateRangeType } from 'config/types';
import { TIMELINE_ITEM_WIDTH } from '../../utils';

interface AAFloodTimelineItemProps {
  currentDate: DateRangeType;
}

function AAFloodTimelineItem({ currentDate }: AAFloodTimelineItemProps) {
  const { availableDates } = useSelector(
    (state: RootState) => state.anticipatoryActionFloodState,
  );
  const [isHovered, setIsHovered] = useState(false);

  // Find the date info for this date
  const dateInfo = availableDates.find(
    d => Math.abs(d.displayDate - currentDate.value) < 24 * 60 * 60 * 1000, // Within 24 hours
  );

  if (!dateInfo) {
    return null;
  }
  const severityColor = dateInfo?.color || '#4CAF50';

  return (
    <div
      style={{
        position: 'absolute',
        height: 24,
        width: TIMELINE_ITEM_WIDTH - 1,
        pointerEvents: 'none',
        top: 0,
        backgroundColor: severityColor,
        borderRadius: '2px',
        transition: 'opacity 0.2s ease',
        opacity: isHovered ? 0.8 : 1,
      }}
      onMouseEnter={() => setIsHovered(true)}
      onMouseLeave={() => setIsHovered(false)}
    />
  );
}

export default AAFloodTimelineItem;
