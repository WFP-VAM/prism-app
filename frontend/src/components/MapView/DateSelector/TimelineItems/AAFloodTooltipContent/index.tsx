import { Typography, Box } from '@material-ui/core';
import { DateRangeType } from 'config/types';
import { useSelector } from 'react-redux';
import { RootState } from 'context/store';
import { formatInUTC } from 'components/MapView/Layers/AnticipatoryActionStormLayer/utils';

interface AAFloodTooltipContentProps {
  date: DateRangeType;
}

function AAFloodTooltipContent({ date }: AAFloodTooltipContentProps) {
  const { availableDates } = useSelector(
    (state: RootState) => state.anticipatoryActionFloodState,
  );

  // Find the date info for the selected date
  const dateInfo = availableDates.find(
    d => Math.abs(d.displayDate - date.value) < 24 * 60 * 60 * 1000, // Within 24 hours
  );

  if (!dateInfo) {
    return (
      <Typography>{formatInUTC(new Date(date.value), 'MM/dd/yy')}</Typography>
    );
  }

  const dateString = new Date(date.value).toLocaleDateString();
  const severityColor = dateInfo?.color || '#4CAF50';

  return (
    <Box style={{ display: 'flex', alignItems: 'center', gap: '8px' }}>
      <Typography>{dateString}</Typography>
      <div
        style={{
          width: '12px',
          height: '12px',
          borderRadius: '50%',
          backgroundColor: severityColor,
        }}
      />
    </Box>
  );
}

export default AAFloodTooltipContent;
