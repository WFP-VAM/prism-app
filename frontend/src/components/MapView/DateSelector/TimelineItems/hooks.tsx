import { AAWindStateReports } from 'context/anticipatoryAction/AAStormStateSlice';
import { useMemo } from 'react';
import { useSelector } from 'react-redux';
import { getDateInUTC } from 'components/MapView/Layers/AnticipatoryActionStormLayer/utils';
import { datesAreEqualWithoutTime } from 'utils/date-utils';
import { WindStateReport } from './types';

const getWindStatesForDate = (
  windStateReports: any,
  date: string | null,
  cycloneName?: string,
) => {
  if (!date || !windStateReports[date]) {
    return [];
  }

  // If cycloneName is provided and exists in the data, return only that cyclone
  if (cycloneName && windStateReports[date]?.[cycloneName]) {
    return [
      {
        states: windStateReports[date][cycloneName],
        cycloneName,
      },
    ];
  }

  // Return array of state objects for all available cyclones
  return Object.entries(windStateReports[date]).map(
    ([cycloneNameTemp, states]) => ({
      states,
      cycloneName: cycloneNameTemp,
    }),
  );
};

export const useWindStatesByTime = (
  currentDate: number,
  cycloneName?: string,
): WindStateReport[] => {
  const windStateReports = useSelector(AAWindStateReports);

  return useMemo(() => {
    const dates = Object.keys(windStateReports);
    if (dates.length === 0) {
      return [];
    }

    // If currentDate is 0 or null, get the latest report
    if (!currentDate) {
      const latestDate = dates.reduce((latest, current) =>
        new Date(current) > new Date(latest) ? current : latest,
      );
      return getWindStatesForDate(windStateReports, latestDate, cycloneName);
    }

    const date = Object.keys(windStateReports).find(analyzedDate => {
      const analyzedDateInUTC = getDateInUTC(analyzedDate, false);
      if (!analyzedDateInUTC) {
        return false;
      }
      const isEqual = datesAreEqualWithoutTime(analyzedDateInUTC, currentDate);
      return isEqual;
    });

    const result = getWindStatesForDate(
      windStateReports,
      date || null,
      cycloneName,
    );
    return result;
  }, [currentDate, windStateReports, cycloneName]);
};
