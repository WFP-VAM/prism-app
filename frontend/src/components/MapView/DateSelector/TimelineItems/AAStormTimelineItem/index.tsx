import { createStyles, makeStyles } from '@material-ui/core';

import 'react-datepicker/dist/react-datepicker.css';
import { DateRangeType } from 'config/types';
import { WindState } from 'prism-common/dist/types/anticipatory-action-storm/windState';
import { TIMELINE_ITEM_WIDTH } from '../../utils';
import { useWindStatesByTime } from '../hooks';

function AAStormTimelineItem({ currentDate }: AAStormTimelineItemProps) {
  const allWindStates = useWindStatesByTime(currentDate.value);
  const classes = useStyles();

  const flattenedWindStates = allWindStates.flatMap(
    windStates => windStates.states,
  );

  const getStylingClass = () => {
    if (flattenedWindStates.length === 0) {
      return classes.emptySpace;
    }

    const states = flattenedWindStates.map(({ state }) => state);

    if (states.includes(WindState.activated_64kt)) {
      return classes.activated2Indicator;
    }
    if (states.includes(WindState.activated_48kt)) {
      return classes.activated1Indicator;
    }
    if (states.includes(WindState.ready)) {
      return classes.readyIndicator;
    }
    return classes.lowRiskIndicator;
  };

  return <div className={getStylingClass()} />;
}

const useStyles = makeStyles(() =>
  createStyles({
    emptySpace: {
      position: 'absolute',
      height: 10,
      width: TIMELINE_ITEM_WIDTH - 1,
      top: 0,
    },
    lowRiskIndicator: {
      position: 'absolute',
      height: 12,
      width: TIMELINE_ITEM_WIDTH - 1,
      pointerEvents: 'none',
      top: 0,
      backgroundColor: '#b5ecf4',
    },
    readyIndicator: {
      position: 'absolute',
      height: 16,
      width: TIMELINE_ITEM_WIDTH - 1,
      pointerEvents: 'none',
      top: 0,
      backgroundColor: '#63B2BD',
    },
    activated1Indicator: {
      position: 'absolute',
      height: 20,
      width: TIMELINE_ITEM_WIDTH - 1,
      pointerEvents: 'none',
      top: 0,
      backgroundColor: '#FF8934',
    },
    activated2Indicator: {
      position: 'absolute',
      height: 24,
      width: TIMELINE_ITEM_WIDTH - 1,
      pointerEvents: 'none',
      top: 0,
      backgroundColor: '#E63701',
    },
  }),
);
interface AAStormTimelineItemProps {
  currentDate: DateRangeType;
}
export default AAStormTimelineItem;
