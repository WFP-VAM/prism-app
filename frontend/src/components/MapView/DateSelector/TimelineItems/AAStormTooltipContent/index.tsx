import { ToggleButton, ToggleButtonGroup } from '@material-ui/lab';
import { DateRangeType } from 'config/types';
import { MouseEvent } from 'react';
import { formatInUTC } from 'components/MapView/Layers/AnticipatoryActionStormLayer/utils';
import { createStyles, makeStyles, Typography } from '@material-ui/core';
import {
  AADataSelector,
  loadStormReport,
  setSelectedStormName,
} from 'context/anticipatoryAction/AAStormStateSlice';
import { useDispatch, useSelector } from 'react-redux';
import { updateDateRange } from 'context/mapStateSlice';
import { getFormattedDate } from 'utils/date-utils';
import { useUrlHistory } from 'utils/url-utils';
import { useWindStatesByTime } from '../hooks';

function AAStormTooltipContent({ date }: AAStormTooltipContentProps) {
  const { updateHistory } = useUrlHistory();
  const stormData = useSelector(AADataSelector);
  const allWindStates = useWindStatesByTime(date.value);
  const classes = useStyles();
  const dispatch = useDispatch();

  const hourToggleHandler = (
    _event: MouseEvent<HTMLElement>,
    value: string,
    cycloneName: string,
  ) => {
    const [, dateValue] = value?.split('::') || ['', ''];
    const stormDate = new Date(dateValue);
    stormDate.setUTCHours(12);
    const time = stormDate.getTime();
    updateHistory('date', getFormattedDate(time, 'default') as string);
    dispatch(updateDateRange({ startDate: time }));
    dispatch(setSelectedStormName(cycloneName));
    dispatch(
      loadStormReport({
        date: dateValue,
        stormName: cycloneName || 'chido',
      }),
    );
  };

  const areMultipleCyclonesActive = allWindStates.length > 1;

  return (
    <div className={classes.container}>
      <div className={classes.dateAndCyclonesContainer}>
        <div className={classes.cyclonesContainer}>
          {allWindStates.map(windStates => (
            <div key={windStates.cycloneName} className={classes.cycloneRow}>
              {areMultipleCyclonesActive && (
                <Typography className={classes.cycloneName}>
                  {windStates.cycloneName?.toUpperCase()}
                </Typography>
              )}
              <ToggleButtonGroup
                value={`${stormData.forecastDetails?.cyclone_name.toUpperCase()}::${stormData.forecastDetails?.reference_time}`}
                exclusive
                onChange={(e, value) =>
                  hourToggleHandler(e, value, windStates.cycloneName || '')
                }
                style={{ marginRight: '8px' }}
              >
                {windStates?.states.map(item => {
                  const itemDate = new Date(item.ref_time);
                  const formattedItemTime = formatInUTC(itemDate, "haaa 'UTC'");

                  return (
                    <ToggleButton
                      key={`${windStates.cycloneName}::${itemDate.valueOf()}`}
                      value={`${windStates.cycloneName?.toUpperCase()}::${item.ref_time}`}
                      onMouseDown={e => e.preventDefault()}
                      className={classes.toggleButton}
                    >
                      <Typography className={classes.time}>
                        {formattedItemTime}
                      </Typography>
                    </ToggleButton>
                  );
                })}
              </ToggleButtonGroup>
            </div>
          ))}
        </div>
        <Typography className={classes.date}>
          {formatInUTC(new Date(date.value), 'MM/dd/yy')}
        </Typography>
      </div>
    </div>
  );
}

const useStyles = makeStyles(() =>
  createStyles({
    container: {
      display: 'flex',
      flexDirection: 'column',
    },
    dateAndCyclonesContainer: {
      display: 'flex',
      alignItems: 'center',
    },
    date: {
      minWidth: 'fit-content',
      gap: '8px',
    },
    cyclonesContainer: {
      display: 'flex',
      flexDirection: 'column',
      gap: '8px',
      width: '100%',
    },
    cycloneRow: {
      display: 'flex',
      alignItems: 'center',
      gap: '8px',
      justifyContent: 'flex-end',
      width: '100%',
    },
    cycloneName: {
      fontSize: '12px',
      fontWeight: 500,
      minWidth: 'fit-content',
    },
    time: {
      fontSize: '12px',
      fontWeight: 400,
      lineHeight: '15px',
      color: '#101010',
      whiteSpace: 'nowrap',
    },
    toggleButton: {
      padding: '6px 6px',
      minHeight: 0,
    },
  }),
);

interface AAStormTooltipContentProps {
  date: DateRangeType;
}

export default AAStormTooltipContent;
