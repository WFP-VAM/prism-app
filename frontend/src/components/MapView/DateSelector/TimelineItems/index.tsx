import { CSSProperties, memo, useCallback, useMemo } from 'react';
import {
  Fade,
  Grid,
  Tooltip,
  createStyles,
  makeStyles,
} from '@material-ui/core';
import { AnticipatoryAction, DateItem, DateRangeType } from 'config/types';
import { grey } from 'muiTheme';
import {
  DateCompatibleLayerWithDateItems,
  TIMELINE_ITEM_WIDTH,
} from 'components/MapView/DateSelector/utils';
import TimelineLabel from './TimelineLabel';
import {
  DARK_BLUE_HEX,
  DARK_GREEN_HEX,
  DARK_ORANGE_HEX,
  LIGHT_BLUE_HEX,
  LIGHT_GREEN_HEX,
  LIGHT_ORANGE_HEX,
} from './utils';
import AAdroughtTooltipContent from './AADroughtTooltipContent';
import AAStormTooltipContent from './AAStormTooltipContent';
import { DateItemStyle } from './types';
import AAStormTimelineItem from './AAStormTimelineItem';
import AADroughtTimelineItem from './AADroughtTimelineItem';

const TimelineItems = memo(
  ({
    dateRange,
    clickDate,
    locale,
    orderedLayers,
    truncatedLayers,
    availableDates,
  }: TimelineItemsProps) => {
    const classes = useStyles();

    // Hard coded styling for date items (first, second, and third layers)
    const DATE_ITEM_STYLING: DateItemStyle[] = useMemo(
      () => [
        {
          class: classes.layerOneDate,
          color: LIGHT_BLUE_HEX,
          layerDirectionClass: classes.layerOneDirection,
          emphasis: classes.layerOneEmphasis,
        },
        {
          class: classes.layerTwoDate,
          color: LIGHT_GREEN_HEX,
          layerDirectionClass: classes.layerTwoDirection,
          emphasis: classes.layerTwoEmphasis,
        },
        {
          class: classes.layerThreeDate,
          color: LIGHT_ORANGE_HEX,
          layerDirectionClass: classes.layerThreeDirection,
          emphasis: classes.layerThreeEmphasis,
        },
        { class: classes.availabilityDate, color: LIGHT_ORANGE_HEX },
      ],
      [classes],
    );

    const isShowingAAStormLayer = orderedLayers.find(
      layer => layer.id === AnticipatoryAction.storm,
    );

    const getTooltipContent = useCallback(
      (date: DateRangeType) =>
        isShowingAAStormLayer ? (
          <AAStormTooltipContent date={date} />
        ) : (
          <AAdroughtTooltipContent
            date={date}
            orderedLayers={orderedLayers}
            dateItemStyling={DATE_ITEM_STYLING}
          />
        ),
      [isShowingAAStormLayer, DATE_ITEM_STYLING, orderedLayers],
    );

    const availableDatesToDisplay = availableDates.filter(
      date => date >= dateRange[0].value,
    );

    // Draw a column for each date of the Timeline that start at the beginning of the year
    return (
      <>
        {dateRange.map((date, index) => {
          const isDateAvailable = availableDatesToDisplay.includes(date.value);
          return (
            <Tooltip
              key={`Root-${date.label}-${date.value}`}
              title={<>{getTooltipContent(date)}</>}
              TransitionComponent={Fade}
              TransitionProps={{ timeout: 0 }}
              placement="top"
              arrow
              {...(isShowingAAStormLayer ? { interactive: true } : null)}
              classes={{
                tooltip: isShowingAAStormLayer
                  ? classes.AAStormTooltip
                  : classes.defaultTooltip,
                arrow: isShowingAAStormLayer
                  ? classes.AAStormTooltipArrow
                  : undefined,
              }}
            >
              <Grid
                item
                xs
                className={`${
                  date.isFirstDay ? classes.dateItemFull : classes.dateItem
                }`}
                onClick={() => clickDate(index)}
                data-date-index={index} // Used by the pointer tick to trigger tooltips
              >
                <TimelineLabel locale={locale} date={date} />

                {isShowingAAStormLayer ? (
                  <AAStormTimelineItem currentDate={date} />
                ) : (
                  <AADroughtTimelineItem
                    concatenatedLayers={truncatedLayers}
                    currentDate={date}
                    dateItemStyling={DATE_ITEM_STYLING}
                    isDateAvailable={isDateAvailable}
                  />
                )}
              </Grid>
            </Tooltip>
          );
        })}
      </>
    );
  },
);

const createLayerStyles = (
  backgroundColor: CSSProperties['backgroundColor'],
  top: CSSProperties['top'],
): LayerStyle => ({
  position: 'absolute',
  height: 10,
  width: TIMELINE_ITEM_WIDTH,
  pointerEvents: 'none',
  opacity: 0.6,
  top,
  backgroundColor,
});

const createDirectionStyles = (
  borderColor: CSSProperties['borderColor'],
  top: CSSProperties['top'],
): DirectionStyle => ({
  top,
  borderLeft: `6px solid ${borderColor}`,
});

const useStyles = makeStyles(() =>
  createStyles({
    dateItemFull: {
      color: '#101010',
      position: 'relative',
      top: -5,
      cursor: 'pointer',
      minWidth: TIMELINE_ITEM_WIDTH,
      '&:hover': {
        borderLeft: '1px solid #101010',
      },
      borderLeft: '1px solid #101010',
      height: 36,
    },
    dateItem: {
      color: '#101010',
      position: 'relative',
      top: -5,
      cursor: 'pointer',
      minWidth: TIMELINE_ITEM_WIDTH,
      '&:hover': {
        borderLeft: '1px solid #101010',
        '& $dayItem': {
          borderLeft: 0,
        },
      },
    },
    defaultTooltip: {
      backgroundColor: '#222222',
      opacity: '0.85 !important',
      maxWidth: 'none',
    },
    AAStormTooltip: {
      backgroundColor: '#FFFFFF',
      border: '1px solid #D3D3D3',
      maxWidth: 'none',
    },
    layerOneDate: createLayerStyles(LIGHT_BLUE_HEX, 0),
    layerTwoDate: createLayerStyles(LIGHT_GREEN_HEX, 10),
    layerThreeDate: createLayerStyles(LIGHT_ORANGE_HEX, 20),
    availabilityDate: createLayerStyles(grey, 0),

    layerOneEmphasis: createLayerStyles(DARK_BLUE_HEX, 0),
    layerTwoEmphasis: createLayerStyles(DARK_GREEN_HEX, 10),
    layerThreeEmphasis: createLayerStyles(DARK_ORANGE_HEX, 20),

    layerOneDirection: createDirectionStyles(DARK_BLUE_HEX, 0),
    layerTwoDirection: createDirectionStyles(DARK_GREEN_HEX, 10),
    layerThreeDirection: createDirectionStyles(DARK_ORANGE_HEX, 20),

    currentDate: {
      border: '2px solid black',
      top: '-7px',
      minWidth: '13.9px',
      maxHeight: '34.05px',
      '&:hover': {
        border: '2px solid black',
      },
    },

    AAStormTooltipArrow: {
      width: '11px',
      height: '11px',
      bottom: '-1px !important',
      '&::before': {
        width: '8px',
        height: '8px',
        backgroundColor: 'white',
        transformOrigin: 'center !important',
        boxSizing: 'border-box',
        borderWidth: '0px 1px 1px 0px',
        borderColor: '#D3D3D3',
        borderStyle: 'solid',
      },
    },
  }),
);

type LayerStyle = {
  position: CSSProperties['position'];
  height: CSSProperties['height'];
  width: CSSProperties['width'];
  pointerEvents: CSSProperties['pointerEvents'];
  opacity: CSSProperties['opacity'];
  top: CSSProperties['top'];
  backgroundColor: CSSProperties['backgroundColor'];
};

type DirectionStyle = {
  top: CSSProperties['top'];
  borderLeft: CSSProperties['borderLeft'];
};

export interface TimelineItemsProps {
  dateRange: DateRangeType[];
  clickDate: (arg: number) => void;
  locale: string;
  availableDates: number[];
  orderedLayers: DateCompatibleLayerWithDateItems[];
  truncatedLayers: DateItem[][];
}

export default TimelineItems;
