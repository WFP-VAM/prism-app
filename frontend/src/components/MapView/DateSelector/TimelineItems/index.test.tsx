import { render } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import TimelineItems from '.';

const props = {
  orderedLayers: [],
  truncatedLayers: [],
  selectedLayerTitles: [],
  availableDates: [],
  dateRange: [
    {
      value: 1640883600000,
      label: '31 Dec 2021',
      month: 'Dec 2021',
      isFirstDay: false,
      date: '2021-12-31',
    },
  ],
  clickDate: () => {},
  locale: 'en',
};

const mockStore = configureStore([]);

const store = mockStore({
  anticipatoryActionDroughtState: {
    availableDates: undefined,
  },
  anticipatoryActionStormState: {
    availableDates: undefined,
  },
  mapState: {
    layers: [],
    dateRange: { startDate: 1678528800000 },
  },
});

test('renders as expected', () => {
  const { container } = render(
    <Provider store={store}>
      <TimelineItems {...props} />
    </Provider>,
  );
  expect(container).toMatchSnapshot();
});
