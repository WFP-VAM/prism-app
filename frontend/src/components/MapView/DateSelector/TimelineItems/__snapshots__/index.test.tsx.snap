// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <mock-tooltip
    arrow="true"
    classes="[object Object]"
    placement="top"
    title="[object Object]"
    transitioncomponent="[object Object]"
    transitionprops="[object Object]"
  >
    <div
      class="MuiGrid-root makeStyles-dateItem-2 MuiGrid-item MuiGrid-grid-xs-true"
      data-date-index="0"
    >
      <div
        class="makeStyles-dayItem-18"
      />
    </div>
  </mock-tooltip>
</div>
`;
